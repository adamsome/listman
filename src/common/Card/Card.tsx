/** @jsx jsx */
import { css, jsx } from '@emotion/core'
import React from 'react'
import FlexBox from '../Flexbox'
import withButtonProps from '../with-button-props'

type Props = typeof defaultProps & {
  onClick?: (event: React.MouseEvent) => void
  children: React.ReactNode
  /**
   * Set to whether the card is first in a collection to manually control
   * borders. `last` must be set as well. If undefined, component will use
   * css selectors (e.g. `first-of-type`) to control borders.
   */
  first?: boolean
  /**
   * Set to whether the card is last in a collection to manually control
   * borders. `first` must be set as well. If undefined, component will use
   * css selectors (e.g. `first-of-type`) to control borders.
   */
  last?: boolean
}

const defaultProps = {}

const Card = (props: Props) => {
  const { first, last, ...rest } = withButtonProps(props)
  const { onClick } = rest

  const manualBorders = first != null && last != null
  const manualFirst = manualBorders && first
  const manualLast = manualBorders && last

  return (
    <FlexBox
      css={css`
        background-color: var(--bg-content);
        border: 1px solid var(--border-content);
        border-top-left-radius: ${manualFirst && 'var(--space-border-radius)'};
        border-top-right-radius: ${manualFirst && 'var(--space-border-radius)'};
        border-bottom-left-radius: ${manualLast &&
          'var(--space-border-radius)'};
        border-bottom-right-radius: ${manualLast &&
          'var(--space-border-radius)'};
        border-bottom-width: ${manualLast ? '1px' : 0};
        &:first-of-type {
          border-top-left-radius: ${!manualBorders &&
            'var(--space-border-radius)'};
          border-top-right-radius: ${!manualBorders &&
            'var(--space-border-radius)'};
        }
        &:last-of-type {
          border-bottom-left-radius: ${!manualBorders &&
            'var(--space-border-radius)'};
          border-bottom-right-radius: ${!manualBorders &&
            'var(--space-border-radius)'};
          border-bottom-width: ${!manualBorders && '1px'};
        }
        color: var(--body);
        cursor: ${onClick ? 'pointer' : 'auto'};
        font-size: var(--font-size-body);
        outline: none;
        overflow: hidden;
        text-align: start;

        &:focus {
          box-shadow: 0 0 0 2px var(--border-focus), var(--shadow);
        }
      `}
      {...rest}
    />
  )
}

Card.defaultProps = defaultProps

export default Card
