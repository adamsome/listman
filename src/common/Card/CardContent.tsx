import styled from '@emotion/styled'
import React from 'react'
import { FlexItem } from '../Flexbox'

type Props = typeof defaultProps & {
  children: React.ReactNode
}

const defaultProps = {}

const CardContent = (props: Props) => {
  const { children, ...rest } = props

  return (
    <StyledFlexItem flex="1" {...rest}>
      <Content>{children}</Content>
    </StyledFlexItem>
  )
}

const StyledFlexItem = styled(FlexItem)`
  margin-right: var(--space-inset);
  padding-top: var(--space-stack);
  padding-bottom: var(--space-stack);
  &:first-of-type {
    margin-left: var(--space-inset);
  }
  overflow: hidden;
`

const Content = styled.div`
  display: block;
`

CardContent.defaultProps = defaultProps

export default CardContent
