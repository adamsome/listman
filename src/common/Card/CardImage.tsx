import styled from '@emotion/styled'
import React from 'react'

type Props = typeof defaultProps &
  React.PropsWithoutRef<JSX.IntrinsicElements['img']> & {}

const defaultProps = {}

const CardImage = (props: Props) => {
  return <Image alt={props.alt || ''} {...props} />
}

const Image = styled.img<Props>`
  display: block;
  max-width: 100%;
  margin-left: var(--space-inset);
  margin-right: var(--space-inset);
  &:first-of-type {
    margin-left: 0;
  }
  &:last-of-type {
    margin-right: 0;
  }
`

CardImage.defaultProps = defaultProps

export default CardImage
