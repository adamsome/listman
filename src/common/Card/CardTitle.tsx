/** @jsx jsx */
import { css, jsx } from '@emotion/core'
import styled from '@emotion/styled'
import React, { Children, cloneElement } from 'react'
import FlexBox, { FlexItem } from '../Flexbox'
import Truncated, { TruncatedProps } from '../Truncated'

type Props = typeof defaultProps & {
  children: React.ReactNode
  subtitle?: string
  subtitleLines?: number
  actions?: React.ReactElement | React.ReactElement[]
}

const defaultProps = {
  lines: 1 as number,
}

const CardTitle = (props: Props) => {
  const { children, subtitle, subtitleLines, actions, lines, ...rest } = props

  const _actions = actions
    ? Children.map(actions, (a, i) => cloneElement(a, { key: i }))
    : null

  return (
    <Wrapper alignItems="flex-start" {...rest}>
      <FlexItem flex="auto" css={titlesWrapper(actions != null)}>
        <Title lines={lines}>{children}</Title>
        {subtitle && <Subtitle lines={subtitleLines || 1}>{subtitle}</Subtitle>}
      </FlexItem>
      <FlexItem flex="none">{_actions}</FlexItem>
    </Wrapper>
  )
}

const Wrapper = styled(FlexBox)`
  &:not(:last-of-type) {
    margin-bottom: var(--space-stack-small);
  }
  &:not(:first-of-type) {
    margin-top: var(--space-stack-small);
  }
`

const titlesWrapper = (hasActions: boolean) => css`
  margin-right: ${hasActions ? '1rem' : null};
  display: block;
  min-width: 0;
`

const StyledTitle = styled.h4`
  color: var(--strong);
  font-size: var(--font-size-big);
  font-weight: var(--font-weight-bold);
  margin: 0;
`

const Title = (props: TruncatedProps) => {
  return (
    <StyledTitle>
      <Truncated {...props} />
    </StyledTitle>
  )
}

const StyledSubtitle = styled(StyledTitle)`
  color: var(--subtle);
`

const Subtitle = (props: TruncatedProps) => {
  return (
    <StyledSubtitle>
      <Truncated {...props} />
    </StyledSubtitle>
  )
}

CardTitle.defaultProps = defaultProps

export default CardTitle
