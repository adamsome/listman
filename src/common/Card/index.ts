export { default } from './Card'
export { default as CardContent } from './CardContent'
export { default as CardImage } from './CardImage'
export { default as CardTitle } from './CardTitle'
