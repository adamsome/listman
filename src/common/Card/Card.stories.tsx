import { action } from '@storybook/addon-actions'
import React from 'react'
import Starrate from 'react-minor-ui'
import Truncated from '../Truncated'
import Card from './Card'
import CardContent from './CardContent'
import CardImage from './CardImage'
import CardTitle from './CardTitle'
import sampleArtImage from './sample-art.png'

export default {
  component: Card,
  title: 'Card',
  excludeStories: ['actions', 'createCard', 'SAMPLE_CARDS'],
}

export const actions = {
  onClick: action('onClick'),
}

const longTitle = 'Long title of an type of artifact that is being rated'
const shortSubtitle = 'Short subtitle'
const longDescription =
  'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Phasellus lobortis rhoncus risus, ut faucibus dui. Aenean bibendum dui et finibus euismod. Nullam luctus nisl at ipsum faucibus, sed volutpat metus ultricies. Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia Curae; Praesent dictum ligula at sollicitudin venenatis. Fusce facilisis tincidunt accumsan. Vestibulum vitae efficitur urna. Maecenas aliquam accumsan scelerisque. Morbi molestie, justo sed pretium gravida, erat nulla mattis turpis, non aliquam sapien orci vel arcu. Curabitur id mauris a odio pharetra varius eget quis ipsum. Donec ipsum tellus, bibendum quis maximus eu, facilisis vitae nibh. Maecenas euismod mauris at orci eleifend efficitur.'

interface SampleCard {
  title: string
  subtitle?: string
  description?: string
  rating?: number
}

export const SAMPLE_CARDS: Record<string, SampleCard> = {
  short: {
    title: 'Title',
    subtitle: 'Subtitle',
    description: 'Description',
  },
  long: {
    title: longTitle,
    subtitle: shortSubtitle,
    description: longDescription,
  },
  noSubtitle: {
    title: 'Title with no subtitle',
    description: longDescription,
  },
  color1: {
    title: 'Red Cow Mountain',
    subtitle: 'Green Bull Lake',
  },
  color2: {
    title: 'Blue Deer Ocean',
    subtitle: 'Yellow Doe Sky',
  },
  color3: {
    title: 'Orange Chicken Plains',
    subtitle: 'Purple Rooster Canyon',
  },
  color4: {
    title: 'Pink Bear Valley',
    subtitle: 'Forest Beaver Hill',
  },
  color5: {
    title: 'Black Lion River',
    subtitle: 'White Tiger Desert',
  },
}

const DEFAULT_SAMPLE: SampleCard = {
  title: longTitle,
  subtitle: shortSubtitle,
  description: 'No description.',
}

export const createCard = (card: Partial<SampleCard> = DEFAULT_SAMPLE) => {
  const { title, subtitle, description, rating } = {
    ...DEFAULT_SAMPLE,
    ...card,
  }
  return (
    <Card {...actions}>
      <CardImage src={sampleArtImage} height="114px" alt="Sample Art" />
      <CardContent>
        <CardTitle subtitle={subtitle} actions={<Starrate rating={rating} />}>
          {title}
        </CardTitle>
        <Truncated lines={2}>{description}</Truncated>
      </CardContent>
    </Card>
  )
}

export const image = () => createCard()
export const lowWidth = () => (
  <div style={{ width: '400px' }}>{createCard()}</div>
)

// TODO: Story w/ long titles
