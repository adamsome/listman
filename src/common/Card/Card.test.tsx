import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import { createCard } from './Card.stories'

it('renders without crashing', () => {
  const div = document.createElement('div')
  ReactDOM.render(createCard(), div)
  ReactDOM.unmountComponentAtNode(div)
})

it('renders correctly', () => {
  const tree = renderer.create(createCard()).toJSON()
  expect(tree).toMatchSnapshot()
})
