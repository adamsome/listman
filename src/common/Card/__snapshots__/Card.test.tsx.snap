// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
.emotion-11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: var(--bg-content);
  border: 1px solid var(--border-content);
  border-bottom-width: 0;
  color: var(--body);
  cursor: pointer;
  font-size: var(--font-size-body);
  outline: none;
  overflow: hidden;
  text-align: start;
}

.emotion-11:first-of-type {
  border-top-left-radius: var(--space-border-radius);
  border-top-right-radius: var(--space-border-radius);
}

.emotion-11:last-of-type {
  border-bottom-left-radius: var(--space-border-radius);
  border-bottom-right-radius: var(--space-border-radius);
  border-bottom-width: 1px;
}

.emotion-11:focus {
  box-shadow: 0 0 0 2px var(--border-focus),var(--shadow);
}

.emotion-0 {
  display: block;
  max-width: 100%;
  margin-left: var(--space-inset);
  margin-right: var(--space-inset);
}

.emotion-0:first-of-type {
  margin-left: 0;
}

.emotion-0:last-of-type {
  margin-right: 0;
}

.emotion-10 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  margin-right: var(--space-inset);
  padding-top: var(--space-stack);
  padding-bottom: var(--space-stack);
  overflow: hidden;
}

.emotion-10:first-of-type {
  margin-left: var(--space-inset);
}

.emotion-9 {
  display: block;
}

.emotion-7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.emotion-7:not(:last-of-type) {
  margin-bottom: var(--space-stack-small);
}

.emotion-7:not(:first-of-type) {
  margin-top: var(--space-stack-small);
}

.emotion-5 {
  -webkit-flex: auto;
  -ms-flex: auto;
  flex: auto;
  margin-right: 1rem;
  display: block;
  min-width: 0;
}

.emotion-2 {
  color: var(--strong);
  font-size: var(--font-size-big);
  font-weight: var(--font-weight-bold);
  margin: 0;
}

.emotion-1 {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.emotion-4 {
  color: var(--strong);
  font-size: var(--font-size-big);
  font-weight: var(--font-weight-bold);
  margin: 0;
  color: var(--subtle);
}

.emotion-6 {
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
}

.emotion-8 {
  display: -webkit-box;
  -webkit-line-clamp: 2;
  -webkit-box-orient: vertical;
  overflow: hidden;
}

<div
  className="emotion-11"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="button"
  tabIndex={0}
>
  <img
    alt="Sample Art"
    className="emotion-0"
    height="114px"
    src="sample-art.png"
  />
  <div
    className="emotion-10"
  >
    <div
      className="emotion-9"
    >
      <div
        className="emotion-7"
      >
        <div
          className="emotion-5"
        >
          <h4
            className="emotion-2"
          >
            <div
              className="emotion-1"
            >
              Long title of an type of artifact that is being rated
            </div>
          </h4>
          <h4
            className="emotion-4"
          >
            <div
              className="emotion-1"
            >
              Short subtitle
            </div>
          </h4>
        </div>
        <div
          className="emotion-6"
        >
          <div>
            <svg
              height="1.5rem"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              style={
                Object {
                  "cursor": "pointer",
                }
              }
              width="1.5rem"
            >
              <g
                style={
                  Object {
                    "opacity": 0.25,
                    "transition": "opacity 100ms ease-in-out",
                  }
                }
                x={-0.2}
                y={0.9}
              >
                <svg
                  viewBox="0 0 512 512"
                >
                  <circle
                    cx="256"
                    cy="256"
                    fill="currentColor"
                    r="40"
                    strokeWidth="0"
                    transform="inherit"
                    transform-origin="center"
                  />
                </svg>
              </g>
              <g
                style={
                  Object {
                    "opacity": 0,
                    "transform": "scale(0)",
                    "transformOrigin": "center",
                    "transition": "opacity 200ms ease-in-out, transform 100ms ease-in-out",
                  }
                }
              >
                <svg
                  viewBox="0 0 51 48"
                >
                  <path
                    d="m25,1 6,17h18l-14,11 5,17-15-10-15,10 5-17-14-11h18z"
                    fill="currentColor"
                    stroke="currentColor"
                  />
                </svg>
              </g>
            </svg>
            <svg
              height="1.5rem"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              style={
                Object {
                  "cursor": "pointer",
                }
              }
              width="1.5rem"
            >
              <g
                style={
                  Object {
                    "opacity": 0.25,
                    "transition": "opacity 100ms ease-in-out",
                  }
                }
                x={-0.2}
                y={0.9}
              >
                <svg
                  viewBox="0 0 512 512"
                >
                  <circle
                    cx="256"
                    cy="256"
                    fill="currentColor"
                    r="40"
                    strokeWidth="0"
                    transform="inherit"
                    transform-origin="center"
                  />
                </svg>
              </g>
              <g
                style={
                  Object {
                    "opacity": 0,
                    "transform": "scale(0)",
                    "transformOrigin": "center",
                    "transition": "opacity 200ms ease-in-out, transform 100ms ease-in-out",
                  }
                }
              >
                <svg
                  viewBox="0 0 51 48"
                >
                  <path
                    d="m25,1 6,17h18l-14,11 5,17-15-10-15,10 5-17-14-11h18z"
                    fill="currentColor"
                    stroke="currentColor"
                  />
                </svg>
              </g>
            </svg>
            <svg
              height="1.5rem"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              style={
                Object {
                  "cursor": "pointer",
                }
              }
              width="1.5rem"
            >
              <g
                style={
                  Object {
                    "opacity": 0.25,
                    "transition": "opacity 100ms ease-in-out",
                  }
                }
                x={-0.2}
                y={0.9}
              >
                <svg
                  viewBox="0 0 512 512"
                >
                  <circle
                    cx="256"
                    cy="256"
                    fill="currentColor"
                    r="40"
                    strokeWidth="0"
                    transform="inherit"
                    transform-origin="center"
                  />
                </svg>
              </g>
              <g
                style={
                  Object {
                    "opacity": 0,
                    "transform": "scale(0)",
                    "transformOrigin": "center",
                    "transition": "opacity 200ms ease-in-out, transform 100ms ease-in-out",
                  }
                }
              >
                <svg
                  viewBox="0 0 51 48"
                >
                  <path
                    d="m25,1 6,17h18l-14,11 5,17-15-10-15,10 5-17-14-11h18z"
                    fill="currentColor"
                    stroke="currentColor"
                  />
                </svg>
              </g>
            </svg>
            <svg
              height="1.5rem"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              style={
                Object {
                  "cursor": "pointer",
                }
              }
              width="1.5rem"
            >
              <g
                style={
                  Object {
                    "opacity": 0.25,
                    "transition": "opacity 100ms ease-in-out",
                  }
                }
                x={-0.2}
                y={0.9}
              >
                <svg
                  viewBox="0 0 512 512"
                >
                  <circle
                    cx="256"
                    cy="256"
                    fill="currentColor"
                    r="40"
                    strokeWidth="0"
                    transform="inherit"
                    transform-origin="center"
                  />
                </svg>
              </g>
              <g
                style={
                  Object {
                    "opacity": 0,
                    "transform": "scale(0)",
                    "transformOrigin": "center",
                    "transition": "opacity 200ms ease-in-out, transform 100ms ease-in-out",
                  }
                }
              >
                <svg
                  viewBox="0 0 51 48"
                >
                  <path
                    d="m25,1 6,17h18l-14,11 5,17-15-10-15,10 5-17-14-11h18z"
                    fill="currentColor"
                    stroke="currentColor"
                  />
                </svg>
              </g>
            </svg>
            <svg
              height="1.5rem"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              style={
                Object {
                  "cursor": "pointer",
                }
              }
              width="1.5rem"
            >
              <g
                style={
                  Object {
                    "opacity": 0.25,
                    "transition": "opacity 100ms ease-in-out",
                  }
                }
                x={-0.2}
                y={0.9}
              >
                <svg
                  viewBox="0 0 512 512"
                >
                  <circle
                    cx="256"
                    cy="256"
                    fill="currentColor"
                    r="40"
                    strokeWidth="0"
                    transform="inherit"
                    transform-origin="center"
                  />
                </svg>
              </g>
              <g
                style={
                  Object {
                    "opacity": 0,
                    "transform": "scale(0)",
                    "transformOrigin": "center",
                    "transition": "opacity 200ms ease-in-out, transform 100ms ease-in-out",
                  }
                }
              >
                <svg
                  viewBox="0 0 51 48"
                >
                  <path
                    d="m25,1 6,17h18l-14,11 5,17-15-10-15,10 5-17-14-11h18z"
                    fill="currentColor"
                    stroke="currentColor"
                  />
                </svg>
              </g>
            </svg>
          </div>
        </div>
      </div>
      <div
        className="emotion-8"
      >
        No description.
      </div>
    </div>
  </div>
</div>
`;
